<a href="https://colab.research.google.com/github/BenShieh233/NLP-Text-Mining/blob/main/IC6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Loading packages and importing libraries

In [5]:
import pandas as pd
import plotly.express as px
from textblob import *
import matplotlib.pyplot as plt

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [20]:
df = pd.read_csv("/content/drive/MyDrive/INFOSYS765/Musical_instruments_reviews.csv")

In [21]:
df.head()

reviewerID        asin  \
0  A2IBPI20UZIR0U  1384719342   
1  A14VAT5EAX3D9S  1384719342   
2  A195EZSQDW3E21  1384719342   
3  A2C00NNG1ZQQG2  1384719342   
4   A94QU4C90B1AX  1384719342   

                                       reviewerName   helpful  \
0  cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1                                              Jake  [13, 14]   
2                     Rick Bennette "Rick Bennette"    [1, 1]   
3                         RustyBill "Sunday Rocker"    [0, 0]   
4                                     SEAN MASLANKA    [0, 0]   

                                          reviewText  overall  \
0  Not much to write about here, but it does exac...        5   
1  The product does exactly as it should and is q...        5   
2  The primary job of this device is to block the...        5   
3  Nice windscreen protects my MXL mic and preven...        5   
4  This pop filter is great. It looks and perform...        5   

                                 summary  unixReviewTime   reviewTime  
0                                   good      1393545600  02 28, 2014  
1                                   Jake      1363392000  03 16, 2013  
2                   It Does The Job Well      1377648000  08 28, 2013  
3          GOOD WINDSCREEN FOR THE MONEY      1392336000  02 14, 2014  
4  No more pops when I record my vocals.      1392940800  02 21, 2014

Dataset preprocessing

In [22]:
df.columns

Index(['reviewerID', 'asin', 'reviewerName', 'helpful', 'reviewText',
       'overall', 'summary', 'unixReviewTime', 'reviewTime'],
      dtype='object')

In [23]:
df = df.rename(columns={'reviewText': 'Review_Text', 'reviewerName': 'reviewer_name'})

In [24]:
df = df[['reviewerID','reviewer_name','Review_Text','overall','summary']]
df.head()

reviewerID                                     reviewer_name   helpful  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1  A14VAT5EAX3D9S                                              Jake  [13, 14]   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"    [1, 1]   
3  A2C00NNG1ZQQG2                         RustyBill "Sunday Rocker"    [0, 0]   
4   A94QU4C90B1AX                                     SEAN MASLANKA    [0, 0]   

                                         Review_Text  overall  \
0  Not much to write about here, but it does exac...        5   
1  The product does exactly as it should and is q...        5   
2  The primary job of this device is to block the...        5   
3  Nice windscreen protects my MXL mic and preven...        5   
4  This pop filter is great. It looks and perform...        5   

                                 summary  
0                                   good  
1                                   Jake  
2                   It Does The Job Well  
3          GOOD WINDSCREEN FOR THE MONEY  
4  No more pops when I record my vocals.

Drop the rows where at least one element is missing

In [25]:
df['Review_Text'].isna().value_counts()

False    10254
True         7
Name: Review_Text, dtype: int64

In [26]:
df = df.dropna()

### (1) Calculate the number of words in each review and store the results in a new column, `text_length`.

In [27]:
df['Text_Length'] = df['Review_Text'].map(lambda text: len(text.split(' ')))
df.head()

reviewerID                                     reviewer_name   helpful  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1  A14VAT5EAX3D9S                                              Jake  [13, 14]   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"    [1, 1]   
3  A2C00NNG1ZQQG2                         RustyBill "Sunday Rocker"    [0, 0]   
4   A94QU4C90B1AX                                     SEAN MASLANKA    [0, 0]   

                                         Review_Text  overall  \
0  Not much to write about here, but it does exac...        5   
1  The product does exactly as it should and is q...        5   
2  The primary job of this device is to block the...        5   
3  Nice windscreen protects my MXL mic and preven...        5   
4  This pop filter is great. It looks and perform...        5   

                                 summary  Text_Length  
0                                   good           51  
1                                   Jake          104  
2                   It Does The Job Well           77  
3          GOOD WINDSCREEN FOR THE MONEY           35  
4  No more pops when I record my vocals.           28

### (2) Calculate the polarity and subjectivity of each review using the package 'TextBlob'. Store the outputs into new columns, `polarity` and `subjectivity`.

In [29]:
df['polarity'] = df['Review_Text'].map(lambda text: TextBlob(text).sentiment.polarity)

In [30]:
df['subjectivity'] = df['Review_Text'].map(lambda text: TextBlob(text).sentiment.subjectivity)

In [31]:
df.head()

reviewerID                                     reviewer_name   helpful  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1  A14VAT5EAX3D9S                                              Jake  [13, 14]   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"    [1, 1]   
3  A2C00NNG1ZQQG2                         RustyBill "Sunday Rocker"    [0, 0]   
4   A94QU4C90B1AX                                     SEAN MASLANKA    [0, 0]   

                                         Review_Text  overall  \
0  Not much to write about here, but it does exac...        5   
1  The product does exactly as it should and is q...        5   
2  The primary job of this device is to block the...        5   
3  Nice windscreen protects my MXL mic and preven...        5   
4  This pop filter is great. It looks and perform...        5   

                                 summary  Text_Length  polarity  subjectivity  
0                                   good           51  0.250000      0.398611  
1                                   Jake          104  0.052778      0.394444  
2                   It Does The Job Well           77  0.167500      0.434286  
3          GOOD WINDSCREEN FOR THE MONEY           35  0.200000      0.925000  
4  No more pops when I record my vocals.           28  0.800000      0.750000

### (3) Calculate the magnitude of the positive, neutral, negative, and compound sentiment of each review using the package 'VaderSentiment'. Store the outputs into new columns, `vader_pos`, `vader_neu`, `vader_neg`, and `vader_compound`.

In [32]:
!pip install vaderSentiment

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.0/126.0 KB 10.8 MB/s eta 0:00:00


In [33]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [35]:
analyzer = SentimentIntensityAnalyzer()

In [36]:
df['vader_pos'] = df['Review_Text'].map(lambda text: analyzer.polarity_scores(text)['pos'])
df['vader_neg'] = df['Review_Text'].map(lambda text: analyzer.polarity_scores(text)['neg'])
df['vader_neu'] = df['Review_Text'].map(lambda text: analyzer.polarity_scores(text)['neu'])
df['vader_compound'] = df['Review_Text'].map(lambda text: analyzer.polarity_scores(text))

In [37]:
df.head()

reviewerID                                     reviewer_name   helpful  \
0  A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1  A14VAT5EAX3D9S                                              Jake  [13, 14]   
2  A195EZSQDW3E21                     Rick Bennette "Rick Bennette"    [1, 1]   
3  A2C00NNG1ZQQG2                         RustyBill "Sunday Rocker"    [0, 0]   
4   A94QU4C90B1AX                                     SEAN MASLANKA    [0, 0]   

                                         Review_Text  overall  \
0  Not much to write about here, but it does exac...        5   
1  The product does exactly as it should and is q...        5   
2  The primary job of this device is to block the...        5   
3  Nice windscreen protects my MXL mic and preven...        5   
4  This pop filter is great. It looks and perform...        5   

                                 summary  Text_Length  polarity  subjectivity  \
0                                   good           51  0.250000      0.398611   
1                                   Jake          104  0.052778      0.394444   
2                   It Does The Job Well           77  0.167500      0.434286   
3          GOOD WINDSCREEN FOR THE MONEY           35  0.200000      0.925000   
4  No more pops when I record my vocals.           28  0.800000      0.750000   

   vader_pos  vader_neg  vader_neu  \
0      0.153      0.057      0.790   
1      0.171      0.000      0.829   
2      0.029      0.110      0.861   
3      0.199      0.055      0.746   
4      0.202      0.000      0.798   

                                      vader_compound  
0  {'neg': 0.057, 'neu': 0.79, 'pos': 0.153, 'com...  
1  {'neg': 0.0, 'neu': 0.829, 'pos': 0.171, 'comp...  
2  {'neg': 0.11, 'neu': 0.861, 'pos': 0.029, 'com...  
3  {'neg': 0.055, 'neu': 0.746, 'pos': 0.199, 'co...  
4  {'neg': 0.0, 'neu': 0.798, 'pos': 0.202, 'comp...

In [57]:
px.histogram(df, x = 'polarity', color_discrete_sequence=['purple'])

### (4) Select the first 100 rows and review each review's content and generated sentiment scores from (2) and (3). Compare to the results from (2) and (3), which package do you think is more effective in sentiment analysis?

In [55]:
df1['Review_Text'][4]

"This pop filter is great. It looks and performs like a studio filter. If you're recording vocals this will eliminate the pops that gets recorded when you sing."

In [53]:
df1 = df.head(100)
for review in df1['Review_Text']:
  print(review)

Not much to write about here, but it does exactly what it's supposed to. filters out the pop sounds. now my recordings are much more crisp. it is one of the lowest prices pop filters on amazon so might as well buy it, they honestly work the same despite their pricing,
The product does exactly as it should and is quite affordable.I did not realized it was double screened until it arrived, so it was even better than I had expected.As an added bonus, one of the screens carries a small hint of the smell of an old grape candy I used to buy, so for reminiscent's sake, I cannot stop putting the pop filter next to my nose and smelling it after recording. :DIf you needed a pop filter, this will work just as well as the expensive ones, and it may even come with a pleasing aroma like mine did!Buy this product! :]
The primary job of this device is to block the breath that would otherwise produce a popping sound, while allowing your voice to pass through with no noticeable reduction of volume or hi

I think both sentiment measures from (2) and (3) are quite accurate. But VaderSentiment seems to place more weights on the neutral sentiments. For example, in the fifth review, TextBlob gives an high polarity score as well as a high subjectivity score, while VaderSentiment has a neutral score of 0.798 and a positive score of 0.202. However, the latter package might work better when there is a combination of sentiments. And in that case, a higher neutral score may be more effective to avoid wrong sentiment analysis.

### (5) Do two more analyses on the Musical Instrument Review Data

(5.1) Apply Word.spellcheck() to the first 100 rows of `Review_Text`

In [ ]:
df1['Review_Text']

In [74]:
from textblob import Word

In [79]:
df1['spellcheck'] = df1['Review_Text'].map(lambda text: [Word(word).spellcheck() for word in text.split(' ')])

<ipython-input-79-ce125a5c8987>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [80]:
df1

reviewerID                                     reviewer_name  \
0    A2IBPI20UZIR0U  cassandra tu "Yeah, well, that's just like, u...   
1    A14VAT5EAX3D9S                                              Jake   
2    A195EZSQDW3E21                     Rick Bennette "Rick Bennette"   
3    A2C00NNG1ZQQG2                         RustyBill "Sunday Rocker"   
4     A94QU4C90B1AX                                     SEAN MASLANKA   
..              ...                                               ...   
96   A3M48SSAOTBSMW                             L. Rapoport "lenrapp"   
97    AFLRU6952DEFX                                                S.   
98   A1W3CEEQBJ4GTN                                        S. Marchuk   
99    AXP9CF1UTFRSU                                              tada   
100  A1IR3NR4KRPIJ1                                       Jason Lewis   

      helpful                                        Review_Text  overall  \
0      [0, 0]  Not much to write about here, but it does exac...        5   
1    [13, 14]  The product does exactly as it should and is q...        5   
2      [1, 1]  The primary job of this device is to block the...        5   
3      [0, 0]  Nice windscreen protects my MXL mic and preven...        5   
4      [0, 0]  This pop filter is great. It looks and perform...        5   
..        ...                                                ...      ...   
96     [3, 3]  Just received this cord and it seems to work a...        5   
97     [0, 0]  If you're like me, you probably bought this to...        4   
98     [0, 0]  I bought this for my Canon Vixia HF G10 Video ...        2   
99     [0, 0]  Well made, XLR 3 pin adaptor to mini.  Needed ...        5   
100    [0, 0]  The listing for this item can be confusing if ...        5   

                                               summary  Text_Length  polarity  \
0                                                 good           51  0.250000   
1                                                 Jake          104  0.052778   
2                                 It Does The Job Well           77  0.167500   
3                        GOOD WINDSCREEN FOR THE MONEY           35  0.200000   
4                No more pops when I record my vocals.           28  0.800000   
..                                                 ...          ...       ...   
96                                   Works as expected           36  0.500000   
97                                  Good for some uses           40  0.263333   
98    Measure your damned wires before sending them...          189  0.069549   
99                                           Recommend           24  0.200000   
100  Product shipd as expected, but the listing for...          104 -0.056473   

     subjectivity  vader_pos  vader_neg  vader_neu  \
0        0.398611      0.153      0.057      0.790   
1        0.394444      0.171      0.000      0.829   
2        0.434286      0.029      0.110      0.861   
3        0.925000      0.199      0.055      0.746   
4        0.750000      0.202      0.000      0.798   
..            ...        ...        ...        ...   
96       0.600000      0.221      0.000      0.779   
97       0.400000      0.102      0.045      0.853   
98       0.517084      0.068      0.070      0.862   
99       0.300000      0.087      0.000      0.913   
100      0.340575      0.000      0.038      0.962   

                                        vader_compound  \
0    {'neg': 0.057, 'neu': 0.79, 'pos': 0.153, 'com...   
1    {'neg': 0.0, 'neu': 0.829, 'pos': 0.171, 'comp...   
2    {'neg': 0.11, 'neu': 0.861, 'pos': 0.029, 'com...   
3    {'neg': 0.055, 'neu': 0.746, 'pos': 0.199, 'co...   
4    {'neg': 0.0, 'neu': 0.798, 'pos': 0.202, 'comp...   
..                                                 ...   
96   {'neg': 0.0, 'neu': 0.779, 'pos': 0.221, 'comp...   
97   {'neg': 0.045, 'neu': 0.853, 'pos': 0.102, 'co...   
98   {'neg': 0.07, 'neu': 0.862, 'pos': 0.068, 'com...   
99   

In [82]:
df1['spellcheck']

0      [[(Not, 0.9338971106412967), (Got, 0.039464411...
1      [[(The, 0.8303848428566982), (He, 0.1286717785...
2      [[(The, 0.8303848428566982), (He, 0.1286717785...
3      [[(Vice, 0.3391812865497076), (Nice, 0.3099415...
4      [[(His, 0.7117826487905228), (This, 0.28821735...
                             ...                        
96     [[(Must, 0.5386350817822899), (Just, 0.4326001...
97     [[(Of, 0.9362354096980188), (If, 0.05550747350...
98     [[(I, 1.0)], [(bought, 1.0)], [(this, 1.0)], [...
99     [[(Well, 0.602918973326623), (Tell, 0.24760946...
100    [[(The, 0.8303848428566982), (He, 0.1286717785...
Name: spellcheck, Length: 100, dtype: object

(5.2) Using Naive Bayes to predict the sentiment score of the first 1000 rows of reviews, given that `sentiment_score` = 1 if `overall` > 3.

In [104]:
df.overall.unique()

array([5, 3, 4, 2, 1])

In [105]:
df['sentiment_score'] = df['overall'].apply(lambda x: 1 if x>3 else 0)
df

index      reviewerID  \
0          0  A2IBPI20UZIR0U   
1          1  A14VAT5EAX3D9S   
2          2  A195EZSQDW3E21   
3          3  A2C00NNG1ZQQG2   
4          4   A94QU4C90B1AX   
...      ...             ...   
10222  10256  A14B2YH83ZXMPP   
10223  10257   A1RPTVW5VEOSI   
10224  10258   AWCJ12KBO5VII   
10225  10259  A2Z7S8B5U4PAKJ   
10226  10260  A2WA8TDCTGUADI   

                                          reviewer_name   helpful  \
0      cassandra tu "Yeah, well, that's just like, u...    [0, 0]   
1                                                  Jake  [13, 14]   
2                         Rick Bennette "Rick Bennette"    [1, 1]   
3                             RustyBill "Sunday Rocker"    [0, 0]   
4                                         SEAN MASLANKA    [0, 0]   
...                                                 ...       ...   
10222                                   Lonnie M. Adams    [0, 0]   
10223                                Michael J. Edelman    [0, 0]   
10224                                  Michael L. Knapp    [0, 0]   
10225                           Rick Langdon "Scriptor"    [0, 0]   
10226                                   TheTerrorBeyond    [0, 0]   

                                             Review_Text  overall  \
0      Not much to write about here, but it does exac...        5   
1      The product does exactly as it should and is q...        5   
2      The primary job of this device is to block the...        5   
3      Nice windscreen protects my MXL mic and preven...        5   
4      This pop filter is great. It looks and perform...        5   
...                                                  ...      ...   
10222            Great, just as expected.  Thank to all.        5   
10223  I've been thinking about trying the Nanoweb st...        5   
10224  I have tried coated strings in the past ( incl...        4   
10225  Well, MADE by Elixir and DEVELOPED with Taylor...        4   
10226  These strings are really quite good, but I wou...        4   

                                                 summary  Text_Length  \
0                                                   good           51   
1                                                   Jake          104   
2                                   It Does The Job Well           77   
3                          GOOD WINDSCREEN FOR THE MONEY           35   
4                  No more pops when I record my vocals.           28   
...                                                  ...          ...   
10222                                         Five Stars            8   
10223  Long life, and for some players, a good econom...          575   
10224                                   Good for coated.          120   
10225                                        Taylor Made          313   
10226  These strings are really quite good, but I wou...          112   

       polarity  subjectivity  vader_pos  vader_neg  vader_neu  \
0      0.250000      0.398611      0.153      0.057      0.790   
1      0.052778      0.394444      0.171      0.000      0.829   
2      0.167500      0.434286      0.029      0.110      0.861   
3      0.200000      0.925000      0.199      0.055      0.746   
4      0.800000      0.750000      0.202      0.000      0.798   
...         ...           ...        ...        ...        ...   
10222  0.350000      0.575000      0.569      0.000      0.431   
10223  0.188033      0.464348      0.139      0.038      0.823   
10224  0.197768      0.401459      0.096      0.043      0.861   
10225  0.153843      0.535669      0.097      0.051      0.852   
10226  0.335333      0.572000      0.175      0.079      0.746   

                                          vader_compound  sentiment_score  
0      {'neg': 0.057, 'neu': 0.79, 'pos': 0.153, 'com...                1  
1      {'neg': 0.0, 'neu': 0.829, 'pos': 0.171, 'comp...                1  
2      {'neg': 0.11, 'neu': 0.861, 'pos': 0.029, 'com...          

In [106]:
df2 = df.head(1000)

In [107]:
import numpy as np
import nltk
import string
import re
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import TfidfVectorizer
nltk.download('stopwords')
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


True

In [108]:
stopwords = nltk.corpus.stopwords.words('english')
wn = WordNetLemmatizer()

def clean_text(text):
    words = "".join([word for word in text if word not in string.punctuation])
    tokens = re.split("\W+", words)
    text = [wn.lemmatize(word) for word in tokens if word not in stopwords]
    return text

In [109]:
countvec = CountVectorizer(analyzer = clean_text)
X_count = countvec.fit_transform(df2["Review_Text"])

df3 = pd.DataFrame(X_count.toarray(), columns=countvec.get_feature_names_out())
df3['label'] = df2['sentiment_score']
df3

016NY  046mm  060mm  06mm  0818  088mm  09  099  0o  ...  yucky  zero  \
0    0      0      0      0     0     0      0   0    0   0  ...      0     0   
1    1      0      0      0     0     0      0   0    0   0  ...      0     0   
2    0      0      0      0     0     0      0   0    0   0  ...      0     0   
3    0      0      0      0     0     0      0   0    0   0  ...      0     0   
4    0      0      0      0     0     0      0   0    0   0  ...      0     0   
..  ..    ...    ...    ...   ...   ...    ...  ..  ...  ..  ...    ...   ...   
995  0      0      0      0     0     0      0   0    0   0  ...      0     0   
996  0      0      0      0     0     0      0   0    0   0  ...      0     0   
997  0      0      0      0     0     0      0   0    0   0  ...      0     0   
998  0      0      0      0     0     0      0   0    0   0  ...      0     0   
999  0      0      0      0     0     0      0   0    0   0  ...      0     0   

     zip  zipper  zippered  ziptie  zone  zoneIf  zoom  label  
0      0       0         0       0     0       0     0      1  
1      0       0         0       0     0       0     0      1  
2      0       0         0       0     0       0     0      1  
3      0       0         0       0     0       0     0      1  
4      0       0         0       0     0       0     0      1  
..   ...     ...       ...     ...   ...     ...   ...    ...  
995    0       0         0       0     0       0     0      1  
996    0       0         0       0     0       0     0      1  
997    0       0         0       0     0       0     0      1  
998    0       0         0       0     0       0     0      1  
999    0       0         0       0     0       0     0      0  

[1000 rows x 7418 columns]

In [110]:
features = df3.drop(['label'], axis = 1)
X_train, X_test, y_train, y_test = train_test_split(features, df3['label'], test_size = 0.2, random_state = 1)

In [111]:
naive_bayes = MultinomialNB()
naive_bayes.fit(X_train, y_train)
y_pred_class = naive_bayes.predict(X_test)

In [112]:
print("Accuracy: ", accuracy_score(y_test, y_pred_class))

Accuracy:  0.87


Therefore, the Naive Bayes has an accuracy of 0.87.